internal/http/parser/http_response.go

```diff
@@ -0,0 +1,263 @@
+				return []Header{}, IncompleteHTTPResponseError{
+					Reader:  reader,
+					Message: "Expected CRLF after header value",
+				}
+			}
+		}
+		if len(possibleHeaderLine) == 0 {
+			// Headers finished
+			allHeadersFound = true
+		} else {
+			// We know header is present
+			reader.Seek(int64(offsetBeforeCRLF), io.SeekStart)
+			key, err := ReadUntil(reader, []byte(":"))
+			if err == io.EOF {
+				reader.Seek(int64(-2), io.SeekCurrent)
+				return []Header{}, IncompleteHTTPResponseError{
+					Reader:  reader,
+					Message: "Expected ':' after header key",
+				}
+			}
+			if key[len(key)-1] == ' ' {
+				// No WhiteSpace before separator
+				return []Header{}, InvalidHTTPResponseError{
+					Reader:  reader,
+					Message: "No whitespace allowed before header separator",
+				}
+			}
+			value, err := ReadUntilCRLF(reader)
+			if err == io.EOF {
+				return []Header{}, IncompleteHTTPResponseError{
+					Reader:  reader,
+					Message: "Expected CRLF after header value",
+				}
+			}
+			header := Header{
+				// 9110: 5.5-5: Replace CR, LF or NUL with SP
+				Key:   string(key),
+				Value: strings.TrimSpace(string(ReplaceCharsWithSpace(value, [][]byte{CR, LF, NUL}))),
+			}
+			headers = append(headers, header)
+		}
+	}
+
+	return headers, nil
+}
+
+func parseContent(reader *bytes.Reader, contentLength int) ([]byte, error) {
+	// If content is present
+	if contentLength != -1 {
+		content, err := ReadBytes(reader, contentLength)
+		if err == io.EOF {
+			return []byte{}, IncompleteHTTPResponseError{
+				Reader:  reader,
+				Message: fmt.Sprintf("Expected content of length %d, Received content of length %d", contentLength, len(content)),
+			}
+		}
+		return content, nil
+	}
+	return []byte{}, nil
+}
+
+func doParseResponse(reader *bytes.Reader) (HTTPResponse, error) {
+	var httpResponse HTTPResponse
+
+	statusLine, err := parseStatusLine(reader)
+	if err != nil {
+		return HTTPResponse{}, err
+	}
+	httpResponse.StatusLine = statusLine
+
+	headers, err := parseHeaders(reader)
+	if err != nil {
+		return HTTPResponse{}, err
+	}
+	httpResponse.Headers = headers
+
+	body, err := parseContent(reader, httpResponse.ContentLength())
+	if err != nil {
+		return HTTPResponse{}, err
+	}
+	httpResponse.Body = body
+
+	return httpResponse, nil
+}
+
+// FindHeader returns a value of a header matching name.
+func (response *HTTPResponse) FindHeader(key string) string {
+	for _, header := range response.Headers {
+		if strings.EqualFold(header.Key, key) {
+			return header.Value
+		}
+	}
+	return ""
+}
+
+// ContentLength returns the value of the Content-Length header.
+// A value of -1 indicates the header was not set.
+func (response *HTTPResponse) ContentLength() int {
+	headerValue := response.FindHeader("Content-Length")
+	if headerValue != "" {
+		contentLength, err := strconv.Atoi(headerValue)
+		if err != nil {
+			return -1
+		}
+		return contentLength
+	}
+
+	return -1
+}
+
+func (response *HTTPResponse) FormattedString() string {
```

Comment:
What is this used for? If we're trying to print out the user's response, then we should just print the raw bytes using `%q` - this can cause confusion since our constructed answer might not match what they sent 1:1.